documentation/porting_guidelines.md

```diff
@@ -0,0 +1,391 @@
+Porting guidelines for OP-TEE
+=============================
+
+1. [Introduction](#1-introduction)
+2. [Add a new platform](#2-add-a-new-platform)
+3. [Hardware Unique Key](#3-hardware-unique-key)
+4. [Secure Clock](#4-secure-clock)
+5. [Root and Chain of Trust](#5-root-and-chain-of-trust)
+6. [Hardware Crypto IP](#6-hardware-crypto-ip)
+7. [Power Management / PSCI](#7-power-management--psci)
+8. [Memory firewalls / TZASC](#8-memory-firewalls--tzasc)
+9. [Trusted Application private/public keypair](#9-trusted-application-privatepublic-keypair)
+
+## 1. Introduction
+---------------
+This document serves a dual purpose, first it should serve as a base for getting
+OP-TEE up and running on a new device and this is the first part of this
+document (section 2). However, even when you have managed to setup the device
+and all tests passes, there are still quite a few things to consider if you
+intend to make a real secure product and that is what the second part of this
+document is about. It shall also be noted that we are trying our best to
+implement full end to end security in OP-TEE in a generic way, but due to the
+nature of devices being different, NDA etc, it's not always possible for us to
+do so and in those cases, we most often try to write a generic API, but we will
+just stub the code. This porting guideline highlights the "gaps", that must be
+addressed in a real secure consumer device. Hopefully we will sooner or later
+get access to devices where we at least can make reference implementations
+publicly available to everyone for the missing pieces we're talking about here.
+
+## 2. Add a new platform
+The first thing you need to do after you have decided to port OP-TEE to another
+device is to add a new platform device. That can either be adding a new platform
+variant (`PLATFORM_FLAVOR`) if it is a device from a family already supported,
+or it can be a brand new platform family (`PLATFORM`). Typically this initial
+setup involve configuring UART, memory addresses etc. For simplicity let's call
+our fictive platform for "gendev" just so we have something to refer to when
+writing examples further down.
+
+### 2.1 core/arch/arm
+In `core/arch/arm` you will find all the currently supported devices. That is
+where you are supposed to add a new platform or modify an existing one.
+Typically you will find this set of files in a specific platform folder:
+```
+$ ls
+conf.mk  kern.ld.S  link.mk  main.c  platform_config.h  sub.mk
+```
+
+##### conf.mk
```

Comment:
It should where this file is supposed to be stored, that is, in `core/arch/arm/plat-gendev`